src/transaction/buffer.rs

```diff
@@ -70,6 +74,56 @@ impl Buffer {
         Ok(results)
     }
 
+    pub async fn scan_and_fetch<F, Fut>(
+        &self,
+        range: BoundRange,
+        limit: u32,
+        f: F,
+    ) -> Result<impl Iterator<Item = KvPair>>
+    where
+        F: FnOnce(BoundRange, u32) -> Fut,
+        Fut: Future<Output = Result<Vec<KvPair>>>,
+    {
+        // read from local buffer
+        let mutations = self.mutations.lock().unwrap();
+        let mutation_range = mutations.range(range.clone());
+
+        // fetch from TiKV
+        // fetch more entries because some of them may be deleted.
+        let redundant_limit = limit
+            + mutation_range
+                .clone()
+                .filter(|(_, m)| matches!(m, Mutation::Del))
+                .count() as u32;
+
+        let mut results = f(range, redundant_limit)
+            .await?
+            .into_iter()
+            .map(|pair| pair.into())
+            .collect::<HashMap<Key, Value>>();
+
+        // override using local data
```

Comment:
For the results which are not overridden, we should insert them into the buffer as `Mutation::Cached`